File: problem/Best_Time/maximumProfit.js
/**
 * @param {number[]} A
 * @param {number} k
 * @return {number}
 */
var maximumProfit = function(A, k) {
    let res = new Array(k + 1).fill(0);
    let bought = new Array(k).fill(-1e9);
    let sold = new Array(k).fill(0);

    for (let a of A) {
        for (let j = k; j >= 1; --j) {
            res[j] = Math.max(res[j], bought[j - 1] + a, sold[j - 1] - a);
            bought[j - 1] = Math.max(bought[j - 1], res[j - 1] - a);
            sold[j - 1] = Math.max(sold[j - 1], res[j - 1] + a);
        }
    }
    return Math.max(...res);
};

File: problem/Best_Time/maximumProfit.c
#include <stdlib.h>
#include <string.h>

long long max2(long long a, long long b)
{
    return a > b ? a : b;
}

long long max3(long long a, long long b, long long c)
{
    return max2(a, max2(b, c));
}

long long maximumProfit(int *A, int ASize, int k)
{
    long long *res = (long long *)calloc(k + 1, sizeof(long long));
    long long *bought = (long long *)malloc(k * sizeof(long long));
    long long *sold = (long long *)calloc(k, sizeof(long long));

    for (int i = 0; i < k; ++i)
    {
        bought[i] = -1000000000LL;
    }

    for (int i = 0; i < ASize; ++i)
    {
        int a = A[i];
        for (int j = k; j >= 1; --j)
        {
            res[j] = max3(res[j], bought[j - 1] + a, sold[j - 1] - a);
            bought[j - 1] = max2(bought[j - 1], res[j - 1] - a);
            sold[j - 1] = max2(sold[j - 1], res[j - 1] + a);
        }
    }

    long long maxVal = res[0];
    for (int i = 1; i <= k; ++i)
    {
        if (res[i] > maxVal)
            maxVal = res[i];
    }

    free(res);
    free(bought);
    free(sold);
    return maxVal;
}